app/services/search/book_versions/indexing_service.rb

```diff
@@ -0,0 +1,51 @@
+module Search::BookVersions
+  class IndexingService
+    DEFAULT_INDEXING_VERSION = I1::IndexMapping
+
+    def initialize(book_guid:,
+                   book_version:,
+                   indexing_mapping: DEFAULT_INDEXING_VERSION)
+      @book_guid = book_guid
+      @book_version = book_version
+      @index_mapping = indexing_mapping.new
+    end
+
+    def create_index
+      ElasticsearchClient.instance.indices.create(index: index_name, body: @index_mapping.index_metadata)
+    end
+
+    def add_book_pages_to_search_index
+      starting = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+      book = Openstax::Book.new(uuid: @book_guid, version: @book_version)
+      book.pages.each do |page|
+        page.indexable_elements.each do |element|
+          ElasticsearchClient.instance.index(index: index_name,
+                                             type: Search::BookVersions::I1::PageElementType.index_type,
+                                             body: element.to_h)
```

Comment:
I think there's maybe a chance for better encapsulation here -- on line 24 we are hardcoding `I1` which should be part of the strategy injected into this generic class.

```ruby
book.pages.each do |page|
  page.elements.each do |element|
    # The indexing strategy (inside this call) would know which elements it cares about and 
    # can ignore the rest e.g. this one could choose to index paragraphs but skip figures.  It
    # would also be able to supply the correct strategy-specific type information, body, etc.
    # Inside this call is even a place where the page order is recorded (so that if the strategy
    # is handed a paragraph, a figure, and then another paragraph, and it skips the figure,
    # it could give a position of `1` and then `2` for the paragraph index documents.
    indexing_strategy.index(client: Elasticsearch.instance, element: element)
  end
end
```

Having the method on `Page` be `elements` instead of `indexable_elements` would also get the "indexing" language out of `Page` and make it just more purely about the page contents.

Reply:
Moving the specific indexing responsibility into a strategy-specific class also makes the `IndexingService`'s methods operate at more of the same level, which is nice.

Reply:
@jpslav i spent yesterday refactoring a lot of this to match up with your and Tom's suggestions.  i'm getting this to work right now....and will then wrap tests around it.  after that we can hook up to the controller plumbing (this i want to chat about) and any other infrastructure logistics.